src/rust/cryptography-x509-validation/src/policy/extension.rs

```diff
@@ -171,14 +161,250 @@ impl<B: CryptoOps> ExtensionPolicy<B> {
+        let key_usage: KeyUsage<'_> = extn.value()?;
+
+        if !key_usage.key_cert_sign() {
+            return Err(ValidationError::Other(
+                "keyUsage.keyCertSign must be asserted in a CA certificate".to_string(),
+            ));
+        }
+
+        Ok(())
+    }
+
+    pub(crate) fn basic_constraints<B: CryptoOps>(
+        _policy: &Policy<'_, B>,
+        _cert: &Certificate<'_>,
+        extn: &Extension<'_>,
+    ) -> Result<(), ValidationError> {
+        let basic_constraints: BasicConstraints = extn.value()?;
+
+        if !basic_constraints.ca {
+            return Err(ValidationError::Other(
+                "basicConstraints.cA must be asserted in a CA certificate".to_string(),
+            ));
+        }
+
+        // NOTE: basicConstraints.pathLength is checked as part of
+        // `Policy::permits_ca`, since we need the current chain building
+        // depth to check it.
+
+        Ok(())
+    }
+
+    pub(crate) fn name_constraints<B: CryptoOps>(
+        _policy: &Policy<'_, B>,
+        _cert: &Certificate<'_>,
+        extn: Option<&Extension<'_>>,
+    ) -> Result<(), ValidationError> {
+        if let Some(extn) = extn {
+            let name_constraints: NameConstraints<'_> = extn.value()?;
+
+            let permitted_subtrees_empty = name_constraints
+                .permitted_subtrees
+                .as_ref()
+                .map_or(true, |pst| pst.unwrap_read().is_empty());
+            let excluded_subtrees_empty = name_constraints
+                .excluded_subtrees
+                .as_ref()
+                .map_or(true, |est| est.unwrap_read().is_empty());
+
+            if permitted_subtrees_empty && excluded_subtrees_empty {
+                return Err(ValidationError::Other(
+                    "nameConstraints must have non-empty permittedSubtrees or excludedSubtrees"
+                        .to_string(),
+                ));
+            }
+
+            // NOTE: Both RFC 5280 and CABF require each `GeneralSubtree`
+            // to have `minimum=0` and `maximum=NULL`, but experimentally
+            // not many validators check for this.
+        }
+
+        Ok(())
+    }
+}
+
+pub(crate) mod common {
+    use cryptography_x509::{
+        certificate::Certificate,
+        extensions::{ExtendedKeyUsage, Extension, SequenceOfAccessDescriptions},
+    };
+
+    use crate::{
+        ops::CryptoOps,
+        policy::{Policy, ValidationError},
+    };
+
+    pub(crate) fn authority_information_access<B: CryptoOps>(
+        _policy: &Policy<'_, B>,
+        _cert: &Certificate<'_>,
+        extn: Option<&Extension<'_>>,
+    ) -> Result<(), ValidationError> {
+        if let Some(extn) = extn {
+            // We don't currently do anything useful with these, but we
+            // do check that they're well-formed.
+            let _: SequenceOfAccessDescriptions<'_> = extn.value()?;
+        }
+
+        Ok(())
+    }
+
+    pub(crate) fn extended_key_usage<B: CryptoOps>(
```

Comment:
We need at least minimal verification of EKU for EEs, to check that it has the serverAuth EKU.